core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -381,10 +381,12 @@ class BrokerLifecycleManager(
   private def sendBrokerRegistration(): Unit = {
     val features = new BrokerRegistrationRequestData.FeatureCollection()
     _supportedFeatures.asScala.foreach {
-      case (name, range) => features.add(new BrokerRegistrationRequestData.Feature().
+      // Do not include features with the range 0-0.
+      case (name, range) if range.max() > 0 => features.add(new BrokerRegistrationRequestData.Feature().
```

Comment:
We should do the same for supported features in ApiVersionResponse in the following code in BrokerFeatures, right? Also, to avoid duplicating the code, perhaps it's better to do the filtering logic in BrokerFeatures?
 
```
  def defaultSupportedFeatures(unstableFeatureVersionsEnabled: Boolean): Features[SupportedVersionRange] = {
    val features = new util.HashMap[String, SupportedVersionRange]()
      features.put(MetadataVersion.FEATURE_NAME,
        new SupportedVersionRange(
          MetadataVersion.MINIMUM_KRAFT_VERSION.featureLevel(),
          if (unstableFeatureVersionsEnabled) {
            MetadataVersion.latestTesting.featureLevel
          } else {
            MetadataVersion.latestProduction.featureLevel
          }))
    PRODUCTION_FEATURES.forEach { feature => features.put(feature.featureName,
          new SupportedVersionRange(0,
            if (unstableFeatureVersionsEnabled) {
              feature.latestTesting
            } else {
              feature.latestProduction
            }))
    }
    Features.supportedFeatures(features)
  }
```

Reply:
I had considered putting it in BrokerFeatures/QuorumFeatures, but decided against it. 
Are you suggesting that I remove how Colin wrote the code (in the ApiVersionsResponse itself).
I was trying to avoid removing his code, but if you think it is better that way I can do it. (I will also need to move the test I wrote elsewhere)

Reply:
The main thing is that we need to exclude range 0-0 for supported features in ApiVersionResponse too. It would be better if exclusion logic is shared between ApiVersionResponse and BrokerRegistrationRequest.

Reply:
Yeah. It is already excluded from Colin's commit. If we want to share the logic we can move it to Broker/QuorumFeatures, but I wasn't sure if there was a reason why Colin did it in the response before. 

cc: @cmccabe

Reply:
Can you be clearer about what you're proposing?

From my point of view:

1. We should never report a `0-0` supported versions range in an RPC. There is no point! This applies to both `ApiVersionsResponse` and `BrokerRegistrationRequest`.

2. Obviously a `1-0` range is even worse, being incorrect as well as useless.

3. Having a HashMap or something that contains a `0-0` range is fine, and may simplify the code in some ways. For example if we have code that changes the maximum supported range based on a config, it would probably be annoying to have to mutate the HashMap or use a different hash map if that config was on or off.

I don't feel really strongly about the third point, but I suspect that the code will be messier if you try to remove features with supported version 0-0 from the internal data structures than if you don't.

Reply:
My proposal was that in defaultSupportedFeatues, we iterate over the features and add them to a map. Instead of just adding them all, we check if the range is 0-0. If so, we don't add them. Thus, we don't need specific logic in the apis to remove the feature later. 

The only concern is if somehow the features map expected a feature to exist. Not sure I follow mutating if the config is on or off. I assume we can't change the config to enable unstable features dynamically. (If we did, we'd have to mutate the map anyway)

Reply:
Pushed code with this example.

Reply:
@jolshan : If you think it's easier to just omit them from the map, that's fine.

I still think we should be checking that these things don't exist later on (and for example, throw an exception if they do) to avoid regressions if someone changes something (as has been happening a lot lately. :) )

Reply:
I'm happy to try to make things safer. Can you point me to an example of where you would expect to check if things exist? Do you mean like when building the request? Or something else?